File: mi_primer_paquete/src/publicador.cpp
#include "ros/ros.h"  // Incluye la biblioteca principal de ROS para C++
#include "geometry_msgs/Twist.h"  // Incluye el mensaje tipo Twist para controlar la velocidad

int main(int argc, char **argv)
{
  // Inicializa el nodo ROS llamado "nodo_publicador" con los argumentos de la línea de comandos
  ros::init(argc, argv, "nodo_publicador");

  // Crea un objeto de tipo NodeHandle, que permite interactuar con el sistema ROS
  ros::NodeHandle n;

  // Crea un publicador que enviará mensajes de tipo geometry_msgs::Twist al tema "turtle1/cmd_vel"
  // El segundo parámetro "10" es el tamaño de la cola del buffer de mensajes
  ros::Publisher cmd_vel_pub = n.advertise<geometry_msgs::Twist>("turtle1/cmd_vel", 10);

  // Define la tasa a la cual se ejecutará el bucle principal (10 Hz)
  ros::Rate loop_rate(10);

  // Crea un mensaje de tipo Twist
  geometry_msgs::Twist cmd_vel_msg;

  // Asigna valores a las componentes de velocidad lineal y angular
  cmd_vel_msg.linear.x = 0.5;  // Velocidad lineal en el eje X (m/s)
  cmd_vel_msg.angular.z = 0.5;  // Velocidad angular en el eje Z (rad/s)

  // Bucle principal que continuará ejecutándose mientras ROS esté en funcionamiento
  while (ros::ok())
  {
    // Imprime un mensaje en la consola para indicar que se está publicando un mensaje de velocidad
    std::cout<< "Publicando velocidad en tortuga"<< std::endl;

    // Publica el mensaje de velocidad en el tema "turtle1/cmd_vel"
    cmd_vel_pub.publish(cmd_vel_msg);
    
    // Pausa el bucle para cumplir con la tasa de actualización de 10 Hz
    loop_rate.sleep();
  }

  // El código nunca llega aquí porque el bucle while es infinito, pero se retorna 0 para indicar ejecución exitosa
  return 0;
}

File: mi_primer_paquete/scripts/publicador.py
#!/usr/bin/env python3
import rospy                              # Importa la librería principal de ROS para Python.
from geometry_msgs.msg import Twist       # Importa el mensaje Twist para enviar comandos de velocidad.

# Función principal del publicador
def publicador():
    rospy.init_node('publicador', anonymous=True)  
    # Inicializa un nodo llamado "publicador".
    # El parámetro `anonymous=True` permite que el nodo tenga un identificador único si se lanzan múltiples instancias.

    pub = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)  
    # Declara un publicador que publicará mensajes en el tópico "/turtle1/cmd_vel".
    # El mensaje que se publicará será de tipo `Twist`, y se usa una cola con un tamaño máximo de 10 mensajes.

    rate = rospy.Rate(10)  
    # Establece la frecuencia de publicación en 10 Hz (10 veces por segundo).

    cmd_vel_msg = Twist()  
    # Crea un mensaje de tipo `Twist` para enviar comandos de velocidad lineal y angular.

    cmd_vel_msg.linear.x = 0.5  
    # Establece la velocidad lineal en el eje X a 0.5 unidades.

    cmd_vel_msg.angular.z = 0.5  
    # Establece la velocidad angular en el eje Z a 0.5 radianes/segundo.

    while not rospy.is_shutdown():  
        # Bucle principal que se ejecutará mientras ROS esté activo y no se haya solicitado una interrupción.

        pub.publish(cmd_vel_msg)  
        # Publica el mensaje `cmd_vel_msg` en el tópico "/turtle1/cmd_vel".

        rospy.loginfo('Publicando')  
        # Imprime un mensaje en la consola indicando que se está publicando información.

        rate.sleep()  
        # Pausa el bucle para mantener la frecuencia de 10 Hz.

# Punto de entrada principal del script
if __name__ == '__main__':
    try:
        publicador()  
        # Llama a la función principal del publicador.
    except rospy.ROSInterruptException:  
        # Captura excepciones si ROS se interrumpe de forma inesperada.
        pass
